bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMappingProvider.java

```diff
@@ -0,0 +1,250 @@
+  private static final String JSON_EXT = ".json";
+  private static String STRUCTURE_DEFINITION = "StructureDefinition";
+
+  /**
+   * This method initially loads all the default base structure definitions into the context and
+   * additionally loads any custom profiles in the given structureDefinitionsPath. It expects only
+   * one extended profile for the given base resource type.
+   *
+   * <p>This method returns a map containing the mappings between the resource type and the profile
+   * url which is defined for the resource type. Any extended profile defined in the
+   * structureDefinitionsPath will overwrite the mapped profile with the extended profile.
+   *
+   * @param context The context to which the profiles are added.
+   * @param structureDefinitionsPath the path containing the list of structure definitions to be
+   *     used
+   * @return the map containing the resource type to profile mapping.
+   * @throws ProfileMapperException if there are any errors while loading and mapping the structure
+   *     definitions
+   */
+  Map<String, String> loadStructureDefinitions(
+      FhirContext context, @Nullable String structureDefinitionsPath, boolean isClasspath)
+      throws ProfileMapperException {
+
+    // TODO: Add support for other versions (R4B and R5) and then remove this constraint
+    // https://github.com/google/fhir-data-pipes/issues/958
+    if (context.getVersion().getVersion() != FhirVersionEnum.DSTU3
+        && context.getVersion().getVersion() != FhirVersionEnum.R4) {
+      String errorMsg =
+          String.format(
+              "Cannot load FHIR profiles for FhirContext version %s as it is not %s or %s ",
+              context.getVersion().getVersion(), FhirVersionEnum.DSTU3, FhirVersionEnum.R4);
+      log.error(errorMsg);
+      throw new ProfileMapperException(errorMsg);
+    }
+
+    PrePopulatedValidationSupport support = new PrePopulatedValidationSupport(context);
+    Map<String, String> baseResourcePofileMap = loadBaseStructureDefinitions(context, support);
+    if (!Strings.isNullOrEmpty(structureDefinitionsPath)) {
+      Map<String, String> customResourceProfileMap =
+          loadCustomStructureDefinitions(context, support, structureDefinitionsPath, isClasspath);
+      // Overwrite the profiles for the resources with the custom profiles
+      baseResourcePofileMap.putAll(customResourceProfileMap);
+    }
+    context.setValidationSupport(support);
+    return baseResourcePofileMap;
+  }
+
+  private Map<String, String> loadBaseStructureDefinitions(
+      FhirContext context, PrePopulatedValidationSupport support) {
+    Map<String, String> resourceProfileMap = new HashMap<>();
+    List<IBaseResource> defaultDefinitions =
+        context.getValidationSupport().fetchAllStructureDefinitions();
+    for (IBaseResource definition : defaultDefinitions) {
+      support.addStructureDefinition(definition);
+      // Links the profile only if the definition belongs to a base resource
```

Comment:
nit: Consider documenting why we need the following check (the same explanation you provide to my question before).

Reply:
Done